[2.1] Add options for more human-friendly build output

timoschwarzer opened the pull request:
This PR adds a little progress indicator to SCons's command line output when building.

The progress indicator is **enabled by default** and can be turned off with `progress=no`.
The options `verbose` and `warnings` now default to `no`.
There is a new quick switch, `dev=yes`, which is an alias for `verbose=yes warnings=all`.

Build output with `progress=yes`:
![image](https://user-images.githubusercontent.com/8849554/27984239-db68e7c0-63d0-11e7-82e3-402c5d6b2a33.png)

timoschwarzer committed:
Add options for more human-friendly build output

eska014 commented:
`human=yes` doesn't make sense with `warnings=no`, humans are the only subjects that would be interested in warnings

timoschwarzer commented:
@eska014 Thought about that, too. But it turned out being pretty hard to find an appropriate name for it.

akien-mga commented:
As a human I feel offended that my favourite option would be `human=no` :p

timoschwarzer commented:
Then let's find a better name and fix the OSX build :smile:

akien-mga commented:
Maybe `simplified_output`.

vnen commented on SConstruct:
```diff
@@ -443,3 +450,40 @@ else:
     print("\nPlease run scons again with argument: platform=<string>")
+
+
+screen = open('/dev/tty', 'w')
```
This isn't cross-platform. It is failing on OSX and will also fail on Windows.

timoschwarzer replied:
Jep, noticed it in the Travis check. Is `sys.stdout` available across the platforms?

vnen replied:
I think it is.

hubbyist commented:
Want about "unattended" to name this feature, since main reason is not to see details of compiling process unless there is something wrong?

akien-mga commented:
After discussion on IRC, here's a proposal:
- make `progress` default to yes
- make `verbose` default to no
- make `warnings` default to no
- add a `dev` option that triggers `verbose=yes warnings=all`, defaults to no

Then no need for a human readable option since that's the default.

timoschwarzer commented:
Great! I will change the commits later today.

timoschwarzer commented:
I've modified and rebased the commits to match the proposal from @akien-mga / IRC (https://github.com/godotengine/godot/pull/9551#issuecomment-314258070)

And I modified the Travis config, too. (`progress=no dev=yes`)

akien-mga commented:
> And I modified the Travis config, too. (progress=no dev=yes)

Actually for Travis it's fine to have the progress IMO, but I would hide the warnings; there are way too many of them and they make the logs huge. So `verbose=yes warnings=no` I guess.

timoschwarzer commented:
@akien-mga The progress indicator only works after completing one build as SCons does count the build nodes only while compiling. This node count gets stored so we can show the progress indicator on the next build. (That's what the `.scons_node_count` file in `.gitignore` is for)
As the Travis build is a clean working directory everytime it can't have a progress indicator.

timoschwarzer commented:
@akien-mga Do you think `warnings=moderate` will be fine for Travis or should I turn them off completely?

akien-mga commented on SConstruct:
```diff
@@ -143,9 +144,11 @@ opts.Add('disable_3d', "Disable 3D nodes for smaller executable (yes/no)", 'no')
-opts.Add('warnings', "Set the level of warnings emitted during compilation (extra/all/moderate/no)", 'all')
+opts.Add('warnings', "Set the level of warnings emitted during compilation (extra/all/moderate/no)", 'no')
+opts.Add('progress', "Show a progress indicator during build (yes/no)", 'yes')
+opts.Add('dev', "If yes, alias for verbose=yes warnings=yes", 'no')
```
Should be "warning=all" in the description ;)

timoschwarzer replied:
Oh, yes :smile:

timoschwarzer replied:
Fixed.

akien-mga commented:
> Do you think warnings=moderate will be fine for Travis or should I turn them off completely?

For now I'd say turn them off completely, Godot produces tons of warnings and we never look at them in Travis logs, so it just makes builds take more time for no gain.

We can change it to moderate once enough warnings have been fixed, so that it actually makes sense to check logs for warnings.

timoschwarzer commented:
@akien-mga Done.

akien-mga commented on SConstruct:
```diff
@@ -11,6 +10,8 @@ import glob
 import sys
 import methods
 
+from SCons.Script import BUILD_TARGETS
```
Is this actually used?

timoschwarzer replied:
Uhm, no. :smile:

akien-mga replied:
Then if you can remove it, it's mergeable :)

akien-mga replied:
BTW from what I could read in the scons docs, I think stuff from SCons.Script is already in scope in SConstruct.

timoschwarzer replied:
@akien-mga Done. And thanks for the hint :)
